scanner/scanner.go

```diff
@@ -0,0 +1,185 @@
+	if scannedSize > size {
+		s.staticLogger.Warnf("Scanned size (%d bytes) is more than the content size (%d bytes) for skylink %s", scannedSize, size, sl.Skylink)
+	}
+	// Once the scanning is complete, we no longer need the skylink. We want to
+	// remove it and only keep its hash. We don't want our database to be an
+	// index of nasty files.
+	sl.Skylink = ""
+	sl.Status = database.SkylinkStatusComplete
+	sl.Infected = inf
+	sl.InfectionDescription = desc
+	sl.Size = size
+	sl.ScannedAllContent = scannedSize == size
+	sl.ScannedAllOffsets = false
+	sl.Timestamp = time.Now().UTC()
+	err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+	if err != nil {
+		s.staticLogger.Debugln(errors.AddContext(err, "updating a skylink's status failed"))
+	}
+	return err
+}
+
+// Start launches a background task that periodically scans the database for
+// new skylink records and sends them for scanning.
+func (s Scanner) Start() {
+	// abort channel which interrupts the current scanning operation
+	abort := make(chan bool, 1)
+	// sleepLength defines how long the thread will sleep before scanning the
+	// next skylink. Its value is controlled by SweepAndScan - while we keep
+	// finding files to scan, we'll keep this sleep at zero. Once we run out of
+	// files to scan we'll reset it to its full duration of sleepBetweenScans.
+	sleepLength := sleepBetweenScans
+	go func() {
+		for {
+			numSubsequentErrs := 0
+			select {
+			case <-s.staticCtx.Done():
+				// interrupt the current scan and exit
+				abort <- true
+				return
+			case <-time.After(sleepLength):
+				err := s.SweepAndScan(abort)
+				// This was a successful call, so the number of subsequent
+				// errors is reset and we sleep for a pre-determined period in
+				// waiting for new skylinks to be uploaded.
+				if errors.Contains(err, database.ErrNoDocumentsFound) {
+					sleepLength = sleepBetweenScans
+					numSubsequentErrs = 0
+					return
+				}
+				// On error, we sleep for an increasing amount of time - from
+				// 100ms on the first error to 100s on the fourth and subsequent
+				// errors.
+				if err != nil {
+					sleepLength = sleepOnErrStep * time.Duration(math.Pow10(numSubsequentErrs))
+					numSubsequentErrs++
+					if numSubsequentErrs > sleepOnErrSteps {
+						numSubsequentErrs = sleepOnErrSteps
+					}
+					return
+				}
+				// A successful scan. Reset the number of subsequent errors.
+				numSubsequentErrs = 0
+				// No need to sleep after a successful scan.
+				sleepLength = 0
+			}
+		}
+	}()
+}
+
+// StartUnlocker launches a background thread that periodically scans the
+// database and resets the state of potentially stuck scans. If a scan has been
+// initiated too long ago it will put it back in "new" state, so it can be
+// retried.
+func (s Scanner) StartUnlocker() {
+	go func() {
+		for {
+			select {
+			case <-s.staticCtx.Done():
+				// interrupt the current scan and exit
+				return
+			case <-time.After(database.ScanTimeout):
```

Comment:
For this I'd just use a `time.NewTicker`. It's slightly better because you don't need to allocate a new timer in the background at every iteration and it will be more accurate.

Reply:
Is this important for the functioning of the service? I am not arguing about the merit of the comment but about its urgency. Can we get this merged and operational, so we can start scanning files and then focus on fine-grain optimisation steps?

Reply:
Ticker added.